packages/redis-cache-nodejs/src/RedisCache.ts

```diff
@@ -0,0 +1,99 @@
+import { AgentContext, Cache, CacheModuleConfig } from '@credo-ts/core'
+import Redis, { RedisOptions } from 'ioredis'
+
+export type RedisCacheOptions = RedisOptions
+
+export class RedisCache implements Cache {
+  private readonly _client: Redis
+
+  constructor(options: RedisCacheOptions = {}) {
+    this._client = new Redis(options)
+  }
+
+  private async client() {
+    try {
+      await this._client.ping()
+      return this._client
+    } catch {
+      await this._client.connect()
+      return this._client
+    }
+  }
+
+  private getNamespacedKey(agentContext: AgentContext, key: string): string {
+    return `${agentContext.contextCorrelationId}:${key}`
+  }
+
+  private serialize<CacheValue>(value: CacheValue): string {
+    return JSON.stringify(value)
+  }
+
+  private deserialize<CacheValue>(value: string | null): CacheValue | null {
+    return value === null ? value : (JSON.parse(value) as CacheValue)
+  }
+
+  private getDefaultExpiryInSeconds(agentContext: AgentContext) {
+    try {
+      return agentContext.resolve(CacheModuleConfig).defaultExpiryInSeconds
+    } catch {
+      return undefined
+    }
+  }
+
+  public async get<CacheValue>(agentContext: AgentContext, key: string): Promise<CacheValue | null> {
+    const client = await this.client()
+    const namespacedKey = this.getNamespacedKey(agentContext, key)
+    const value = await client.get(namespacedKey)
+    return this.deserialize<CacheValue>(value)
+  }
+
+  public async set<CacheValue>(
+    agentContext: AgentContext,
+    key: string,
+    value: CacheValue,
+    expiresInSeconds: number | undefined = this.getDefaultExpiryInSeconds(agentContext)
+  ): Promise<void> {
+    const client = await this.client()
+    const namespacedKey = this.getNamespacedKey(agentContext, key)
+    const serializedValue = this.serialize(value)
+
+    if (expiresInSeconds) {
+      await client.set(namespacedKey, serializedValue, 'EX', expiresInSeconds)
+    } else {
+      await client.set(namespacedKey, serializedValue)
+    }
+  }
+
+  public async remove(agentContext: AgentContext, key: string): Promise<void> {
+    const client = await this.client()
+    const namespacedKey = this.getNamespacedKey(agentContext, key)
+    await client.del(namespacedKey)
+  }
+
+  public async destroy(agentContext: AgentContext) {
```

Comment:
Is this part of the cache interface? or for testing?

Reply:
Becuase it won't work for multi-tenancy cases i think?

Reply:
`destroy` might be a bad name, but disconnects and removes all the tenants key/values.

Reply:
but if it also disconnects, that impacts more than just the tenant? There could be other processes currently doing a query that will then suddenly be aborted?

Reply:
The client checks with a ping if it is still alive. It's less efficient, though.  

I can make it delete the records, but not disconnect,

Reply:
Yes but that's before the query starts. I'm not sure what the behviour is if you call disconnect in the middile of a query while waiting for a response?

I want to avoid this:

- Start destroy
- start remove keys command
- Start request 1
- Request 1 pings
- Request 1 send requests, waiting for response
- remove keys comlete, destory session
- Error because session closed while request 1 was processing

Reply:
But not disconnecting will solve it